File: loops/loop0/events.py
"""Runtime event model used by loop0 observers and components."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4


@dataclass(frozen=True)
class AgentEvent:
    """A structured event emitted during one Agent run."""

    type: str
    run_id: str
    payload: dict[str, Any] = field(default_factory=dict)
    event_id: str = field(default_factory=lambda: f"evt_{uuid4().hex[:12]}")
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

File: src/loop0/events.rs
use chrono::Utc;
use serde_json::Value;
use uuid::Uuid;

use crate::loop0::types::AgentEvent;

pub fn agent_event(event_type: impl Into<String>, run_id: &str, payload: Value) -> AgentEvent {
    AgentEvent {
        r#type: event_type.into(),
        run_id: run_id.to_string(),
        payload,
        event_id: format!("evt_{}", Uuid::new_v4().simple()),
        timestamp: Utc::now(),
    }
}
